Core/src/org/sleuthkit/autopsy/contentviewers/ContextViewer.java

```diff
@@ -0,0 +1,427 @@
+
+                //save the artifact for "Go to Result" button
+                sourceContextArtifact = associatedArtifact;
+
+                setSourceFields(associatedArtifact);
+            }
+        }
+    }
+
+    /**
+     * Sets the source label and text fields based on the given associated
+     * artifact.
+     *
+     * @param associatedArtifact - associated artifact
+     *
+     * @throws TskCoreException
+     */
+    private void setSourceFields(BlackboardArtifact associatedArtifact) throws TskCoreException {
+        if (BlackboardArtifact.ARTIFACT_TYPE.TSK_MESSAGE.getTypeID() == associatedArtifact.getArtifactTypeID()
+                || BlackboardArtifact.ARTIFACT_TYPE.TSK_EMAIL_MSG.getTypeID() == associatedArtifact.getArtifactTypeID()) {
+
+            setSourceName(Bundle.ContextViewer_attachmentSource());
+            setSourceText(msgArtifactToAbbreiviatedString(associatedArtifact));
+
+        } else if (BlackboardArtifact.ARTIFACT_TYPE.TSK_WEB_DOWNLOAD.getTypeID() == associatedArtifact.getArtifactTypeID()
+                || BlackboardArtifact.ARTIFACT_TYPE.TSK_WEB_CACHE.getTypeID() == associatedArtifact.getArtifactTypeID()) {
+
+            setSourceName(Bundle.ContextViewer_downloadSource());
+            setSourceText(webDownloadArtifactToString(associatedArtifact));
+        }
+    }
+
+    /**
+     * Sets the source label string.
+     *
+     * @param nameLabel String value for source label.
+     */
+    private void setSourceName(String nameLabel) {
+        jSourceNameLabel.setText(nameLabel);
+    }
+
+    /**
+     * Sets the source text string.
+     *
+     * @param nameLabel String value for source text.
+     */
+    private void setSourceText(String text) {
+        jSourceTextLabel.setText(text);
+        showSourceText(true);
+    }
+
+    private void showSourceText(boolean isVisible) {
+        jSourceTextLabel.setVisible(isVisible);
+    }
+
+    /**
+     * Returns a display string with download source URL from the given
+     * artifact.
+     *
+     * @param artifact artifact to get download source URL from.
+     *
+     * @return Display string with download URL and date/time.
+     *
+     * @throws TskCoreException
+     */
+    private String webDownloadArtifactToString(BlackboardArtifact artifact) throws TskCoreException {
+        StringBuilder sb = new StringBuilder(1024);
+        Map<BlackboardAttribute.ATTRIBUTE_TYPE, BlackboardAttribute> attributesMap = getAttributesMap(artifact);
+
+        if (BlackboardArtifact.ARTIFACT_TYPE.TSK_WEB_DOWNLOAD.getTypeID() == artifact.getArtifactTypeID()
+                || BlackboardArtifact.ARTIFACT_TYPE.TSK_WEB_CACHE.getTypeID() == artifact.getArtifactTypeID()) {
+            appendAttributeString(sb, BlackboardAttribute.ATTRIBUTE_TYPE.TSK_URL, attributesMap, "URL");
+            appendAttributeString(sb, BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME_CREATED, attributesMap, "On");
```

Comment:
Should "On" and "URL" be defined as messages?  Or if they won't be internationalized, static constants?